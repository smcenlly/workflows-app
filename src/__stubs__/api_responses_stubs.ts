export const apiResponsesStubs = {
    getActivities: [
        {
            url: 'https://dev-api.toladata.io/api/workflowlevel1/578/',
            id: 578,
            name: '0.0.93',
            start_date: '2019-01-02T12:00:00+01:00',
            end_date: '2019-01-19T12:00:00+01:00',
            create_date: '2019-01-29T14:37:31.800032+01:00',
            edit_date: '2019-01-29T14:38:22.160861+01:00',
            workflowlevel1: "https://dev-api.toladata.io/api/workflowlevel1/471/"
        },
        {
            url: 'https://dev-api.toladata.io/api/workflowlevel1/630/',
            id: 630,
            name: 'any_name',
            start_date: null,
            end_date: null,
            create_date: '2019-02-04T20:36:21.987120+01:00',
            edit_date: '2019-02-04T20:36:21.987133+01:00',
            workflowlevel1: "https://dev-api.toladata.io/api/workflowlevel1/23/"
        },
    ],

    addActivity: {
        url: "https://dev-api.toladata.io/api/workflowlevel2/1083/",
        workflowlevel1: "https://dev-api.toladata.io/api/workflowlevel1/5/",
        name: '0.0.93',
        id: 1083,
        start_date: '2019-01-02T12:00:00+01:00',
        end_date: '2019-01-19T12:00:00+01:00',
        create_date: '2019-01-29T14:37:31.800032+01:00',
        edit_date: '2019-01-29T14:38:22.160861+01:00',
    },
};
