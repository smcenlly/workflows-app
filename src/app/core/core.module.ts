import { NgModule, Optional, SkipSelf } from '@angular/core';
import { CommonModule, APP_BASE_HREF } from '@angular/common';
import { throwIfAlreadyLoaded } from './module-import-guard';
import { HTTP_INTERCEPTORS } from '@angular/common/http';
import { RouterModule } from '@angular/router';
import { NavbarComponent } from './components/navbar/navbar.component';
import { AuthInterceptor } from './http-interceptors/authentication-interceptor';

@NgModule({
    imports: [CommonModule, RouterModule],
    exports: [NavbarComponent],
    declarations: [NavbarComponent],
    providers: [{ provide: HTTP_INTERCEPTORS, useClass: AuthInterceptor, multi: true }],
})
export class CoreModule {
    constructor(@Optional() @SkipSelf() parentModule: CoreModule) {
        throwIfAlreadyLoaded(parentModule, 'CoreModule');
    }
}
