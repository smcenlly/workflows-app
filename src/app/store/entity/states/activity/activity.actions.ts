import { Activity } from 'src/app/activities/models/Activity';

export class DeleteActivity {
    static readonly type = '[Activity] Delete Activity';
    constructor(public activityId: number) {}
}
export class AddActivity {
    static readonly type = '[Activity] Add Activity';
    constructor(public activity: Activity, public programId: number) {}
}
export class EditActivity {
    static readonly type = '[Activity] Edit Activity';
    constructor(public activityId: number, public activity: Activity, public programId: number) {}
}
