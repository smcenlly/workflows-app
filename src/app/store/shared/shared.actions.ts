export class FetchData {
    static readonly type = '[Shared] [API] Fetch Data';
}

export class ChangeProgramPage {
    static readonly type = '[Shared] [UI] [Programs] Change page';
    constructor(public pageNumber: number) {}
}
export class ChangeActivitiesPage {
    static readonly type = '[Shared] [UI] [Activities] Change page';
    constructor(public programId: number, public pageNumber: number) {}
}
