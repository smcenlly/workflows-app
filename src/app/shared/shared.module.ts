import { RouterModule } from '@angular/router';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { HttpClientModule } from '@angular/common/http';
import { CommonModule } from '@angular/common';
import { MatSnackBarModule } from '@angular/material/snack-bar';
import { NgProgressModule } from '@ngx-progressbar/core';
import { PaginationModule } from 'ngx-bootstrap/pagination';
import { AppInputsModule } from './components/ui-inputs/inputs.module';
import { TitleComponent } from './components/title/title.component';
import { NgProgressHttpModule } from '@ngx-progressbar/http';

@NgModule({
    imports: [
        CommonModule,
        FormsModule,
        ReactiveFormsModule,
        HttpClientModule,
        RouterModule,
        MatSnackBarModule,
        NgProgressModule,
        AppInputsModule,
        PaginationModule.forRoot(),
        NgProgressModule.forRoot(),
        NgProgressHttpModule.forRoot(),
    ],
    providers: [],
    exports: [
        CommonModule,
        HttpClientModule,
        FormsModule,
        ReactiveFormsModule,
        RouterModule,
        NgProgressModule,
        AppInputsModule,
        PaginationModule,
        TitleComponent,
        MatSnackBarModule
    ],
    declarations: [TitleComponent],
})
export class SharedModule {}
