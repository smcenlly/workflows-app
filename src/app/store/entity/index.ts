import { State } from '@ngxs/store';
import { DictionaryState } from './states/dictionary/dictionary.state';
import { UserState } from './states/user/user.state';
import { ProgramState } from './states/program/program.state';
import { ActivityState } from './states/activity/activity.state';

export const EntityStates = [DictionaryState, UserState, ProgramState, ActivityState];

@State({
    name: 'entityStateModule',
    children: EntityStates,
})

export class EntityStateModule {}
