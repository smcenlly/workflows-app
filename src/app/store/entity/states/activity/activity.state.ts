import { State, Action, StateContext, Selector, createSelector, Store } from '@ngxs/store';
import { tap, catchError } from 'rxjs/operators';

import { ApiService } from 'src/app/core/services/api.service';
import { addProgramIdProp, addWorkflowLevel1Prop } from 'src/app/core/services/utility';
import { of } from 'rxjs';
import { SnackBarService } from 'src/app/core/services/snackbar.service';
import { Activity } from 'src/app/activities/models/Activity';
import { SharedState, SharedStateModel } from 'src/app/store/shared/shared.state';
import { FetchData } from 'src/app/store/shared/shared.actions';
import { DeleteActivity, AddActivity, EditActivity } from './activity.actions';

export interface ActivityStateModel {
    items: Activity[];
}

@State<ActivityStateModel>({
    name: 'activity',
    defaults: {
        items: [],
    },
})
export class ActivityState {
    constructor(private api: ApiService, private sb: SnackBarService) {}

    @Selector()
    public static getState(state: ActivityStateModel): ActivityStateModel {
        return state;
    }

    @Selector()
    static selectActivityByActivityId(state: ActivityStateModel) {
        return (activityId: number) => state.items.find((activity: Activity) => activity.id === activityId);
    }

    @Selector()
    static selectProgramActivities(state: ActivityStateModel) {
        return (programId: number) => {
            return state.items.filter((activity: Activity) => activity.programId === programId);
        };
    }

    @Selector([SharedState])
    static selectTenActivities(state: ActivityStateModel, sharedState: SharedStateModel) {
        return (programId: number) => {
            const pageNumber = SharedState.selectAtivitiesPageNumber(sharedState)(programId);
            const firstIndex = (pageNumber - 1) * 10;
            return this.selectProgramActivities(state)(programId).slice(firstIndex, firstIndex + 10);
        };
    }

    @Selector()
    static selectActivitiesCount(state: ActivityStateModel) {
        return (programId: number) => this.selectProgramActivities(state)(programId).length;
    }

    @Action(FetchData)
    fetchData(ctx: StateContext<ActivityStateModel>) {
        return this.api.getActivities().pipe(
            tap((a: Activity[]) => {
                ctx.setState({
                    items: a.map(activity => addProgramIdProp(activity)),
                });
            }),
            catchError(err => {
                this.sb.emitErrorSnackBar();
                return of([]);
            }),
        );
    }

    @Action(DeleteActivity)
    deleteActivity(ctx: StateContext<ActivityStateModel>, action: DeleteActivity) {
        return this.api.deleteActivity(action.activityId).pipe(
            tap(() => {
                const state = ctx.getState();
                ctx.patchState({
                    items: state.items.filter(activity => activity.id !== action.activityId),
                });
            }),
        );
    }

    @Action(AddActivity)
    addActivity(ctx: StateContext<ActivityStateModel>, action: AddActivity) {
        return this.api.addActivity(addWorkflowLevel1Prop(action.activity, action.programId)).pipe(
            tap((activity: Activity) => {
                const state = ctx.getState();
                ctx.patchState({
                    items: [...state.items, addProgramIdProp(activity)],
                });
            }),
        );
    }

    @Action(EditActivity)
    editActivity(ctx: StateContext<ActivityStateModel>, action: EditActivity) {
        return this.api.updateActivity(action.activityId, addWorkflowLevel1Prop(action.activity, action.programId)).pipe(
            tap((activity: Activity) => {
                const state = ctx.getState();
                const index = state.items.findIndex(activity => activity.id === action.activityId);
                const activities = [...state.items];
                activities[index] = addProgramIdProp(activity);
                ctx.patchState({
                    items: activities,
                });
            }),
        );
    }
}
