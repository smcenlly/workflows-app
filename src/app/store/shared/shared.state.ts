import { State, Action, StateContext, Selector } from '@ngxs/store';

import { ChangeProgramPage, ChangeActivitiesPage } from './shared.actions';

export interface SharedStateModel {
    programsPageNumber: number;
    activitiesPageNumber: { [s: string]: number };
}

@State<SharedStateModel>({
    name: 'shared',
    defaults: {
        programsPageNumber: 1,
        activitiesPageNumber: {},
    },
})
export class SharedState {
    @Selector()
    public static getState(state: SharedStateModel) {
        return state;
    }

    @Selector()
    static selectAtivitiesPageNumber(state: SharedStateModel) {
        return (programId: number) => {
            return state.activitiesPageNumber[programId] || 1;
        };
    }

    @Action(ChangeProgramPage)
    changeProgramPage(ctx: StateContext<SharedStateModel>, action: ChangeProgramPage) {
        ctx.patchState({
            programsPageNumber: action.pageNumber,
        });
    }

    @Action(ChangeActivitiesPage)
    changeActivitiesPage(ctx: StateContext<SharedStateModel>, action: ChangeActivitiesPage) {
        const state = ctx.getState();
        ctx.patchState({
            activitiesPageNumber: {
                ...state.activitiesPageNumber,
                [action.programId]: action.pageNumber,
            },
        });
    }
}
