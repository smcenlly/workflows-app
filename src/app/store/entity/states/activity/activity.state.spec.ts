import { TestBed, async } from '@angular/core/testing';
import { NgxsModule, Store } from '@ngxs/store';
import { ApiService } from 'src/app/core/services/api.service';
import { ActivityState, ActivityStateModel } from './activity.state';
import { FetchData } from 'src/app/store/shared/shared.actions';
import { apiResponsesStubs } from 'src/__stubs__/api_responses_stubs';
import { of } from 'rxjs';
import { SnackBarService } from 'src/app/core/services/snackbar.service';
import { AddActivity } from './activity.actions';

describe('Activity store', () => {
    let store: Store;
    let apiServiceSpy: Partial<ApiService> = {};

    beforeEach(async(() => {
        TestBed.configureTestingModule({
            imports: [NgxsModule.forRoot([ActivityState])],
            providers: [
                { provide: ApiService, useValue: apiServiceSpy },
                { provide: SnackBarService, useValue: { emitSuccessSnackBar() { }, emitErrorSnackBar() { } } }

            ],
        }).compileComponents();
        store = TestBed.get(Store);
        apiServiceSpy = TestBed.get(ApiService);
    }));

    describe('fetchData', () => {
        it('should have the right info in store', async(() => {
            apiServiceSpy.getActivities = jest.fn(() => of(apiResponsesStubs.getActivities));
            store.dispatch(new FetchData());
            expect(store.selectSnapshot(ActivityState.getState)).toMatchSnapshot()
        }));
    });

    describe('add item', () => {
        it('should have the right info in store', async(() => {
            apiServiceSpy.addActivity = jest.fn(() => of(apiResponsesStubs.addActivity));
            store.dispatch(new AddActivity({
                id: null,
                expected_end_date: "2019-03-03T20:38:29.014Z",
                expected_start_date: "2019-03-03T20:38:29.014Z",
                name: "adggh",
                workflowlevel1: "https://dev-api.toladata.io/api/workflowlevel1/5/",
            }, 5));
            expect(store.selectSnapshot(ActivityState.getState)).toMatchSnapshot()
        }));
    });


    


});
