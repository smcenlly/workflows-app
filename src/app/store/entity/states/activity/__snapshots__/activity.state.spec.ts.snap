// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Activity store add item should have the right info in store 1`] = `
Object {
  "items": Array [
    Object {
      "create_date": "2019-01-29T14:37:31.800032+01:00",
      "edit_date": "2019-01-29T14:38:22.160861+01:00",
      "end_date": "2019-01-19T12:00:00+01:00",
      "id": 1083,
      "name": "0.0.93",
      "programId": 5,
      "start_date": "2019-01-02T12:00:00+01:00",
      "url": "https://dev-api.toladata.io/api/workflowlevel2/1083/",
      "workflowlevel1": "https://dev-api.toladata.io/api/workflowlevel1/5/",
    },
  ],
}
`;

exports[`Activity store fetchData should have the right info in store 1`] = `
Object {
  "items": Array [
    Object {
      "create_date": "2019-01-29T14:37:31.800032+01:00",
      "edit_date": "2019-01-29T14:38:22.160861+01:00",
      "end_date": "2019-01-19T12:00:00+01:00",
      "id": 578,
      "name": "0.0.93",
      "programId": 471,
      "start_date": "2019-01-02T12:00:00+01:00",
      "url": "https://dev-api.toladata.io/api/workflowlevel1/578/",
      "workflowlevel1": "https://dev-api.toladata.io/api/workflowlevel1/471/",
    },
    Object {
      "create_date": "2019-02-04T20:36:21.987120+01:00",
      "edit_date": "2019-02-04T20:36:21.987133+01:00",
      "end_date": null,
      "id": 630,
      "name": "any_name",
      "programId": 23,
      "start_date": null,
      "url": "https://dev-api.toladata.io/api/workflowlevel1/630/",
      "workflowlevel1": "https://dev-api.toladata.io/api/workflowlevel1/23/",
    },
  ],
}
`;

exports[`Activity store fetchData should have the right info in store 2`] = `
Object {
  "items": Array [
    Object {
      "create_date": "2019-01-29T14:37:31.800032+01:00",
      "edit_date": "2019-01-29T14:38:22.160861+01:00",
      "end_date": "2019-01-19T12:00:00+01:00",
      "id": 578,
      "name": "0.0.93",
      "programId": 471,
      "start_date": "2019-01-02T12:00:00+01:00",
      "url": "https://dev-api.toladata.io/api/workflowlevel1/578/",
      "workflowlevel1": "https://dev-api.toladata.io/api/workflowlevel1/471/",
    },
    Object {
      "create_date": "2019-02-04T20:36:21.987120+01:00",
      "edit_date": "2019-02-04T20:36:21.987133+01:00",
      "end_date": null,
      "id": 630,
      "name": "any_name",
      "programId": 23,
      "start_date": null,
      "url": "https://dev-api.toladata.io/api/workflowlevel1/630/",
      "workflowlevel1": "https://dev-api.toladata.io/api/workflowlevel1/23/",
    },
  ],
}
`;
